extensions/http-curl/tests/TestServer.h

```diff
@@ -24,65 +24,87 @@
-  mg_init_library(0);
-}
-
-
-CivetServer * start_webserver(std::string &port, std::string &rooturi, CivetHandler *handler, struct mg_callbacks *callbacks, std::string &cert, std::string &ca_cert) {
-  const char *options[] = { "document_root", ".", "listening_ports", port.c_str(), "error_log_file",
-      "error.log", "ssl_certificate", ca_cert.c_str(), "ssl_protocol_version", "4", "ssl_cipher_list",
-      "ALL", "request_timeout_ms", "10000", "enable_auth_domain_check", "no", "ssl_verify_peer", "no", 0 };
-// ECDH+AESGCM+AES256:!aNULL:!MD5:!DSS
-  std::vector<std::string> cpp_options;
-  for (size_t i = 0; i < (sizeof(options) / sizeof(options[0]) - 1); i++) {
-    cpp_options.push_back(options[i]);
-  }
-
-  if (!mg_check_feature(2)) {
-      fprintf(stderr,
-              "Error: Embedded example built with SSL support, "
-              "but civetweb library build without.\n");
-      return 0;
+/**
+ * A wrapper around CivetServer which notifies Handlers before shutdown,
+ * so they wouldn't get stuck (if a handler returns after shutdown is
+ * initiated it might get stuck inside worker_thread_run > consume_socket)
+ */
+class TestServer{
+  struct CivetLibrary{
+    CivetLibrary() {
+      if (getCounter()++ == 0) {
+        mg_init_library(0);
+      }
+    }
+    ~CivetLibrary() {
+      if (--getCounter() == 0) {
+        mg_exit_library();
+      }
+    }
+   private:
+    std::atomic<int>& getCounter() {
+      static std::atomic<int> counter{0};
+      return counter;
+    }
+  };
+ public:
+  TestServer(std::string &port, std::string &rooturi, CivetHandler *handler, struct mg_callbacks *callbacks, std::string &cert, std::string &ca_cert) {
+    const char *options[] = { "document_root", ".", "listening_ports", port.c_str(), "error_log_file",
+                              "error.log", "ssl_certificate", ca_cert.c_str(), "ssl_protocol_version", "4", "ssl_cipher_list",
+                              "ALL", "request_timeout_ms", "10000", "enable_auth_domain_check", "no", "ssl_verify_peer", "no", 0 };
+    // ECDH+AESGCM+AES256:!aNULL:!MD5:!DSS
+    std::vector<std::string> cpp_options;
+    for (size_t i = 0; i < (sizeof(options) / sizeof(options[0]) - 1); i++) {
+      cpp_options.emplace_back(options[i]);
     }
 
+    if (!mg_check_feature(2)) {
+      throw std::runtime_error("Error: Embedded example built with SSL support, "
+                               "but civetweb library build without.\n");
+    }
 
-  //mg_init_library(MG_FEATURES_SSL);
 
-  CivetServer *server = new CivetServer(cpp_options, (CivetCallbacks*)callbacks);
+    //mg_init_library(MG_FEATURES_SSL);
 
-  server->addHandler(rooturi, handler);
+    server_ = utils::make_unique<CivetServer>(cpp_options, (CivetCallbacks*)callbacks);
 
-  return server;
+    addHandler(rooturi, handler);
+  }
 
-}
+  TestServer(std::string &port, std::string &rooturi, CivetHandler *handler) {
+    const char *options[] = {"document_root", ".", "listening_ports", port.c_str(), 0};
```

Comment:
I know this is old code, but as far as I see we could simply pass and std::vector<std::string> to Civetweb's ctor, so this could be something like:
```
std::vector<std::string> options = {"document_root", ".", "listening_ports", port};
server_ = utils::make_unique<CivetServer>(options);
```

The same applies to the above.

Reply:
done